include/cufinufft/utils.h

```diff
@@ -146,6 +137,76 @@ static __forceinline__ __device__ void atomicAddComplexGlobal(
   }
 }
 
+template<typename T> auto arrayrange(int n, T *a, cudaStream_t stream) {
+  const auto d_min_max = thrust::minmax_element(thrust::cuda::par.on(stream), a, a + n);
+
+  // copy d_min and d_max to host
+  T min{}, max{};
+  checkCudaErrors(cudaMemcpyAsync(&min, thrust::raw_pointer_cast(d_min_max.first),
+                                  sizeof(T), cudaMemcpyDeviceToHost, stream));
+  checkCudaErrors(cudaMemcpyAsync(&max, thrust::raw_pointer_cast(d_min_max.second),
+                                  sizeof(T), cudaMemcpyDeviceToHost, stream));
+  return std::make_tuple(min, max);
+}
+
+// Writes out w = half-width and c = center of an interval enclosing all a[n]'s
+// Only chooses a nonzero center if this increases w by less than fraction
+// ARRAYWIDCEN_GROWFRAC defined in defs.h.
+// This prevents rephasings which don't grow nf by much. 6/8/17
+// If n==0, w and c are not finite.
+template<typename T> auto arraywidcen(int n, T *a, cudaStream_t stream) {
+  const auto [lo, hi] = arrayrange(n, a, stream);
+  auto w              = (hi - lo) / 2;
+  auto c              = (hi + lo) / 2;
+  if (std::abs(c) < ARRAYWIDCEN_GROWFRAC * w) {
+    w += std::abs(c);
+    c = 0.0;
+  }
+  return std::make_tuple(w, c);
+}
+
+template<typename T>
```

Comment:
Because then cufinufft will have to build and link against finufft. I do not think is worth bringing finufft as a dependency of cufinufft as it is. Once finufft is de-macroized it will be worth grouping common functionalities in a library that is imported by both.